rfcs/2020-04-04-2221-kubernetes-integration.md

```diff
@@ -0,0 +1,253 @@
+
+This RFC outlines how the Vector will integration with Kubernetes (k8s).
+
+**Note: This RFC is retroactive and meant to seve as an audit to complete our
+Kubernetes integration. At the time of writing this RFC, Vector has already made
+considerable progress on it's Kubernetes integration. It has a `kubernetes`
+source, `kubernetes_pod_metadata` transform, an example daemonset file, and the
+ability automatically reload configuration when it changes. The fundamental
+pieces are mostly in place to complete this integration, but as we approach
+the finish line we're being faced with deeper questions that heavily affect the
+UX. Such as how to properly deploy Vector and exclude it's own logs ([pr#2188]).
+We had planned to perform a 3rd party audit on the integration before
+announcement and we've decided to align this RFC with that process.**
+
+## Motivation
+
+Kubernetes is arguably the most popular container orchestration framework at
+the time of writing this RFC; many large companies, with large production
+deployments, depend heavily on Kubernetes. Kubernetes handles log collection
+but does not facilitate shipping. Shipping is meant to be delegated to tools
+like Vector. This is precisely the use case that Vector was built for. So,
+the motivation is three-fold:
+
+1. A Kubernetes integration is essential to achieving Vector's vision of being
+   the dominant, single collector for observability data.
+2. This will inherently attract large, valuable users to Vector since Kubernetes
+   is generally used with large deployments.
+3. It is currently the #1 requested feature of Vector.
+
+## Guide-level Proposal
+
+**Note: This guide largely follows the format of our existing guides
+([example][guide_example]). There are two perspectives to our guides: 1) A new
+user coming from Google 2) A user that is familar with Vector. This guide is
+from perspective 2.**
+
+This guide covers integrating Vector with Kubernetes. We'll touch on the basic
+concepts of deploying Vector into Kubernetes and walk through our recommended
+[strategy](#strategy). By the end of this guide you'll have a single,
+lightweight, ultra-fast, and reliable data collector ready to ship your
+Kubernetes logs and metrics to any destination you please.
+
+### Strategy
+
+#### How This Guide Works
+
+Our recommended strategy deploys Vector as a Kubernetes [daemonset]. This is
+the most efficient means of collecting Kubernetes observability data since
+Vector is guaranteed to deploy _once_ on each of your Pods. In addition,
+we'll use the [`kubernetes_pod_metadata` transform][kubernetes_pod_metadata_transform]
+to enrich your logs with Kubernetes context. This transform interacts with
+the Kubernetes watch API to collect cluster metadata and update in real-time
+when things change. The following diagram demonstrates how this works:
+
+TODO: insert diagram
+
+### What We'll Accomplish
+
+* Collect data from each of your Kubernetes Pods
```

Comment:
This is what I'd call "Kubernetes logs collection for Cluster Operators". More specifically, this would mean collecting logs from the node's host OS.
Noting, that we have a seemingly related [outstanding question](https://github.com/timberio/vector/blob/k8s-rfc/rfcs/2020-04-04-2221-kubernetes-integration.md#outstanding-questions) 9, but this is a different concern.

So, while nothing prevents users from actually configuring vector to process these logs, I don't think it's worth it for us to do anything extra there to provide explicit support for that.

1. It should already work with manual configuration.
2. Cluster operators mostly know what they're doing, and are capable to configure vector as they require.
3. There's a myriad of configurations we'd have to support, and it'd be very hard (if even possible) to come up with sane defaults.
4. Related to the point above, even with sane defaults, in 95% on cases, cluster operators would want to tailor the configuration for their use case.

What we can do, though, is provide guides, blog posts and explainers with concrete examples for Vector usage for Kubernetes Cluster Operators.

I'll include this into the RFC.

Reply:
> What we can do, though, is provide guides, blog posts and explainers with concrete examples for Vector usage for Kubernetes Cluster Operators.

Yeah, totally agree. A guide is a good place for this kind of info.

Reply:
Added this to the RFC